split `ParserGridDistribution` into `hpp/cpp` file

psychocoderHPC opened the pull request:
hopefully, fix #2898

- Split the grid parser into an interface file `hpp` and the implementation `cpp` to avoid that boost regex is parsed by nvcc. Boost regex is very often the root of compile issues because nvcc is not able to split the host and device part correctly.
- Define `abortWithError` static to allow the usage in multiple object files without creating the linker error ` multiple definitions of ...
- remove `boost::regex` by `std::regex`
- remove `boost::lexical_cast`

psychocoderHPC commented:
@ax3l The suggested refactoring to c++ regex you mentioned https://github.com/ComputationalRadiationPhysics/picongpu/issues/2898#issuecomment-465128305 can be added in a follow-up PR.

ax3l commented:
@psychocoderHPC thx! If you ask me, add the C++11 regex in the same PR. We have to backport both anyway.

psychocoderHPC commented:
I do not think that this PR fill fix #2899 fully. I will also have a look to the resource manager.

sbastrakov commented:
@psychocoderHPC is it not possible to make `abortWithError` inline and non-static?

ax3l commented:
`inline` or `static` for global functions only instruct the linker to either always take the local symbol or assume they are all the same. either way, it will work here:

https://stackoverflow.com/questions/185624/static-variables-in-an-inlined-function

ax3l commented:
> I do not think that this PR fill fix #2899 fully. I will also have a look to the resource manager.

Yep, I just would like to have the `gridDist` parser changes in one place :)

sbastrakov commented:
@ax3l sure, for our case both `static` and `inline` would do the job. Actually, from a purist point of view it is even better to put a function in an anonymous namespace (yes, even in a header file). But practically (even for me) seems like too much typing for a single function.

ax3l commented:
Agreed, an anonymous namespace is actually the best option.

psychocoderHPC commented:
I updated the code and removed the boost dependencies for the parser.
Please be not confused by the strange escaping e.g. ` R"(\{.*})",`. 
Only `{` must be escaped (this behavior cost me over 1h of my lifetime)

psychocoderHPC commented:
This PR is still WIP because I am checking if we can remove the dependency to `boost::regex` now.

ax3l commented:
It seems that `boost::iostreams` and `boost::log` depend on `boost::regex`. `boost::program_options` does surprisingly not.

https://gitlab.kitware.com/cmake/cmake/blob/v3.13.4/Modules/FindBoost.cmake#L845-878

ax3l commented on include/picongpu/initialization/ParserGridDistribution.cpp:
```diff
@@ -25,8 +25,7 @@
 
-#include <boost/regex.hpp>
-#include <boost/lexical_cast.hpp>
+#include <regex>
```
bye, byte GCC 4.8 :stars: 
(`INSTALL.md` already lists GCC 4.9+, so that's fine.)

psychocoderHPC commented:
> It seems that `boost::iostreams` and `boost::log` depend on `boost::regex`. `boost::program_options` does surprisingly not, not even transitively.
> 
> https://gitlab.kitware.com/cmake/cmake/blob/v3.13.4/Modules/FindBoost.cmake#L845-878
> 
> https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.4.3/INSTALL.rst#boost
> https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.4.3/include/picongpu/CMakeLists.txt#L168-L169
> https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.4.3/include/pmacc/PMaccConfig.cmake#L308

@ax3l We do not use boost::log or boost::iostream, I tested now also if we can remove the dependency and it looks good.

psychocoderHPC commented:
@ax3l I removed in the last commit the requirement to have boost::regex. But since splash2txt still need it I only marked boost::regex in the documentation as optional.

I will put on my todo list to remove boost::regex from splash2txt too.

ax3l commented:
looks good to me, runtime tests were successful?

psychocoderHPC commented:
Yes I tested different distributions.
